features/config/dao/impl/src/main/java/org/opennms/features/config/dao/impl/util/XsdModelConverter.java

```diff
@@ -0,0 +1,349 @@
+                ConfigItem configItemForArray = new ConfigItem();
+                configItemForArray.setName(currentConfigItem.getName());
+                configItemForArray.setSchemaRef(xmlSchemaElement.getQName().toString());
+                configItemForArray.setType(ConfigItem.Type.ARRAY);
+                configItemForArray.getChildren().add(currentConfigItem);
+                configItemForParent.getChildren().add(configItemForArray);
+            } else {
+                // There may only be 1 element, add it directly to the parent as an object
+                configItemForParent.getChildren().add(currentConfigItem);
+            }
+        }
+    }
+
+    private ConfigItem getConfigItemForParentElement() {
+        if (configItemStack.size() < 2) {
+            return null;
+        }
+        return configItemStack.get(configItemStack.size() - 2);
+    }
+
+    @Override
+    public void onExitElement(XmlSchemaElement xmlSchemaElement, XmlSchemaTypeInfo xmlSchemaTypeInfo, boolean b) {
+        configItemStack.remove(configItemStack.size() - 1);
+    }
+
+    @Override
+    public void onVisitAttribute(XmlSchemaElement xmlSchemaElement, XmlSchemaAttrInfo xmlSchemaAttrInfo) {
+        currentConfigItem.getChildren().add(getConfigItemForAttribute(xmlSchemaAttrInfo));
+        // Check current type -> needs to be an object to create children
+        // Special case where parent is a simple type with an attribute
+        if (currentConfigItem.isPrimitiveType()) {
+            // Make a duplicate of the current item for the primitive type; current then becomes an object with children
+            ConfigItem child = new ConfigItem();
+            // special logic to handle Xml Value to attribute name mapping
+            String bodyName = this.elementNameToValueNameMap.get(currentConfigItem.getName());
+            child.setName(bodyName != null ? bodyName: currentConfigItem.getName());
+            child.setType(currentConfigItem.getType());
+            child.setSchemaRef(currentConfigItem.getSchemaRef());
+            child.setRequired(currentConfigItem.isRequired());
+            currentConfigItem.setType(ConfigItem.Type.OBJECT);
+            currentConfigItem.setSchemaRef("");
+            currentConfigItem.getChildren().add(child);
+        }
+    }
+
+    private void handleStringRestrictions(ConfigItem item, HashMap<XmlSchemaRestriction.Type, List<XmlSchemaRestriction>> facets) {
+        List<XmlSchemaRestriction> patternFacets = facets.get(XmlSchemaRestriction.Type.PATTERN);
+        if (patternFacets != null && !patternFacets.isEmpty()) {
+            XmlSchemaRestriction restriction = patternFacets.get(0);
+            Object obj = restriction.getValue();
+            if (obj instanceof String) {
+                item.setPattern((String) obj);
+            }
+        }
+    }
+
+    private void handleOtherRestrictions(ConfigItem item, HashMap<XmlSchemaRestriction.Type, List<XmlSchemaRestriction>> facets) {
```

Comment:
The enumeration facet may also be worth supporting.

Reply:
Pattern facets do also apply on non-string values. They are applied to the (normalized) lexical representation of values. Also note, that multiple pattern facets are allowed.